Make build-and-analyze work with stable/20230725 branch of clang

rniwa opened the pull request:
#### 54f8c161d3391ccb10743c3c0cff6eb801c5c9df
<pre>
Make build-and-analyze work with stable/20230725 branch of clang
<a href="https://bugs.webkit.org/show_bug.cgi?id=274546">https://bugs.webkit.org/show_bug.cgi?id=274546</a>

Reviewed by David Kilzer.

Add two new options to build-and-analyze script: --preprocessor-additions to specify
additional #define macros, and --dry-run to print out the &quot;make&quot; command to execute
instead of executing it.

Introduce new macro CLANG_WEBKIT_BRANCH which, if defined, enables [[suppress]] on
class declaration and avoid defining HAVE_SWIFT_CPP_INTEROP so that we can run
the static analysis for the entire WebKit repository using:
<a href="https://github.com/rniwa/llvm-project/tree/webkit">https://github.com/rniwa/llvm-project/tree/webkit</a>

* Source/WTF/wtf/Compiler.h:
* Tools/Scripts/build-and-analyze:
(main):
(parse_args):

Canonical link: <a href="https://commits.webkit.org/279212@main">https://commits.webkit.org/279212@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7fd1144c60da09257303342b3754963948966890

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32097 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5244 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56038 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3483 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55065 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3205 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42831 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2246 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54858 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/29862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45543 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23941 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2853 "Found 1 new test failure: imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1642 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46114 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57630 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52273 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27900 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3001 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50229 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29121 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49506 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30037 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64577 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7747 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28873 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12257 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Tools/Scripts/build-and-analyze:
```diff
@@ -139,12 +139,19 @@ def main(args):
     if args.analyzer_path:
         make_command.extend(['CC={}'.format(analyzer_path), 'CPLUSPLUS={}'.format(analyzer_path)])
+    if args.preprocessor_additions:
+        make_command.extend(['GCC_PREPROCESSOR_ADDITIONS={}'.format(args.preprocessor_additions)])
```
Do we need double-quotes around `'GCC_PREPROCESSOR_ADDITIONS="{}"'` in case there are more than one?

rniwa replied:
So I tried that but that results in the second macro be interpreted as the value of the first macro.
e.g. `A=1 B=1` would be interpreted as `#define A 1 B=1` instead of `#define A 1` and `#define B 1`.

ddkilzer reviewed (approved):
r=me

rniwa committed:
Make build-and-analyze work with stable/20230725 branch of clang
https://bugs.webkit.org/show_bug.cgi?id=274546

Reviewed by David Kilzer.

Add two new options to build-and-analyze script: --preprocessor-additions to specify
additional #define macros, and --dry-run to print out the "make" command to execute
instead of executing it.

Introduce new macro CLANG_WEBKIT_BRANCH which, if defined, enables [[suppress]] on
class declaration and avoid defining HAVE_SWIFT_CPP_INTEROP so that we can run
the static analysis for the entire WebKit repository using:
https://github.com/rniwa/llvm-project/tree/webkit

* Source/WTF/wtf/Compiler.h:
* Tools/Scripts/build-and-analyze:
(main):
(parse_args):

Canonical link: https://commits.webkit.org/279212@main
